llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -761,5 +761,138 @@ getDoubleGRFSplitter(ModuleDesc &&MD, bool EmitOnlyKernelsAsEntryPoints) {
     return std::make_unique<ModuleCopier>(std::move(MD), std::move(Groups));
 }
 
+namespace {
+  struct KernelProperties {
+    KernelProperties() = default;
+
+    KernelProperties(const Function *F) {
+      if (const MDNode *MDN = F->getMetadata("sycl_used_aspects")) {
+        auto ExtractIntegerFromMDNodeOperand = [=](const MDNode *N,
+                                                   unsigned OpNo) -> auto {
+          Constant *C =
+              cast<ConstantAsMetadata>(N->getOperand(OpNo).get())->getValue();
+          return C->getUniqueInteger().getSExtValue();
+        };
+
+        for (size_t I = 0, E = MDN->getNumOperands(); I < E; ++I) {
+          Aspects.insert(ExtractIntegerFromMDNodeOperand(MDN, I));
+        }
+
+      }
+    }
+
+    std::string getName(StringRef BaseName) const {
+      if (Aspects.empty())
+        return BaseName.str() + "-no-aspects";
+
+      std::string Ret = BaseName.str() + "-aspects";
+      for (int A : Aspects) {
+        Ret += "-" + std::to_string(A);
+      }
+      return Ret;
+    }
+
+    SmallSet<int, 4> Aspects;
```

Comment:
Maybe SmallSetVector works?

Reply:
Actually, my thumbs up is premature.
`(Small)SetVector` iterator guarantees iteration in the insertion order. That's not what we need. While operator `==` will work with any insertion and iteration order, `hash` needs sorted order, not insertion order - meeting `1` first then `2` will give the same property set as `2` then `1`.

Reply:
I thought about this and since we don't modify this container here like ever (we only construct it once), it should be enough to simply have a vector + sort it in constructor. Presumably, amount of aspects won't be any huge to cause any significant slowdowns.

Reply:
When processing kernel `A` we may encounter property `1` first then property `2`, so they will be recorded as `{1, 2}` in the Aspects for this kernel,  and for kernel `B` we may encounter properties in different order and record as `{ 2, 1 }`.
So this function
```cpp
unsigned hash() const {
  llvm::hash_code AspectsHash =
          llvm::hash_combine_range(Aspects.begin(), Aspects.end());
  return static_cast<unsigned>(llvm::hash_combine(AspectsHash));
}
```
will return different result for A and B. Am I missing something?

Reply:
> will return different result for A and B

Yeah, that is the problem which existing `SmallSet`. My idea is to use `SmallVector` and have it sorted in `KernelProperties` constructor. This way, `hash` will be the same regardless of order in which aspects were discovered.